Get secret/configmaps information from config for all IdPs

stlaz opened the pull request:
Grabs identity providers' configmaps and secrets from the
top-level config and stores their mapping to volumes/volumeMounts
so that these can later be added to the deployment running OSIN

stlaz commented:
Note to self: get rid of `idpSyncData`, it does not seem to be bringing any value, only overhead

ericavonb commented on pkg/operator2/configsync.go:
```diff
@@ -91,64 +91,189 @@ type sourceData struct {
+	ret.src = secretName
+	ret.path = path
+	ret.volume = volume
+	ret.mount = mount
```
Nit: just do 
```go
ret := sourceData{
	src: secretName,
	path: path,
	volume: volume,
	mount: mount,
}
```

ericavonb commented on pkg/operator2/configsync.go:
```diff
@@ -91,64 +91,189 @@ type sourceData struct {
+	ret.src = cmName
+	ret.path = path
+	ret.volume = volume
+	ret.mount = mount
```
Same nit about removing var and mutations in favor of setting in new struct directly

ericavonb commented on pkg/operator2/configsync.go:
```diff
@@ -91,64 +91,189 @@ type sourceData struct {
+
 		default:
-			panic("TODO")
+			return nil // TODO: some erroring
```
`return fmt.Errorf(“unknown IdentityProvider type: %s”, pc.Type)`

ericavonb replied:
And thank you, no panics!

stlaz replied:
The function currently only returns `[]idpSyncData`, I'd maybe fix this when gets merged with `convertProviderConfigToOsinBytes()`

ericavonb replied:
icky, ok. for another pr

ericavonb commented on pkg/operator2/configsync.go:
```diff
@@ -91,64 +91,189 @@ type sourceData struct {
+			clientCertDest, clientCertData := newSourceDataIDPSecret(i, clientCert, corev1.TLSCertKey)
+
+			clientKey := p.TLSClientKey.Name
+			clientKeyDest, clienKeyData := newSourceDataIDPSecret(i, clientKey, corev1.TLSCertKey)
```
Should be `corev1.TLSPrivateKeyKey`

stlaz replied:
Good catch, I was afraid something might have gotten lost in those copy-pastes.

ericavonb commented on pkg/operator2/configsync.go:
```diff
@@ -91,64 +91,189 @@ type sourceData struct {
+		case configv1.IdentityProviderTypeBasicAuth:
+			p := pc.BasicAuth
+
+			ca := p.CA.Name
```
nit: unnecessary var declaration. Use `p.CA.Name` directly. Same for `clientCert` and `clientKey`

stlaz commented:
Fixed comments, rebased on master.

I also added a refactoring to `convertToData()` which now only returns a pointer to the `idpSyncData` structure. Lets' hope nothing breaks!

ericavonb commented on pkg/operator2/configsync.go:
```diff
@@ -91,67 +90,133 @@ type sourceData struct {
+			secrets[clientCertDest] = clientCertData
+
+			clientKeyDest, clienKeyData := newSourceDataIDPSecret(i, p.TLSClientKey.Name, corev1.TLSPrivateKeyKey)
+			secrets[clientKeyDest] = clienKeyData
```
typo - `s/clienKeyData/clientKeyData/` (missing the 't' in 'client')

stlaz commented:
typo fixed

stlaz commented:
latest push fixes writing to `nil` maps

stlaz commented:
I chipped away the refactoring so that we can merge this earlier. See https://github.com/openshift/cluster-osin-operator/pull/26 for the follow-up.

stlaz commented:
Fixed bug in `newSourceDataIDPConfigMap()`, it had `secretVolume()` instead of the proper volume func.

stlaz commented:
/test e2e-aws

enj commented:
/retest

stlaz commented:
/retest

enj commented:
Tested with 4.0 cluster, htpasswd still works so this hopefully fine.

/retest

stlaz commented:
/retest

stlaz commented:
/retest

enj commented:
/retest

enj commented:
/retest

enj commented:
/retest

enj commented:
/retest

stlaz commented:
/retest

enj commented:
/refresh

stlaz commented:
/test e2e-aws-operator

enj commented:
/retest

stlaz commented:
/retest

enj commented:
`kube:admin` still works.  Time for the :hammer:
